docs/user-guide/node-selection/README.md

```diff
@@ -81,14 +81,96 @@ metadata:
+There are currently two types of node affinity, called `requiredDuringSchedulingIgnoredDuringExecution` and
+`preferresDuringSchedulingIgnoredDuringExecution`. You can think of them as "hard" and "soft" respectively,
+in the sense that the former specifies rules that *must* be met for a pod to schedule onto a node (just like
+`nodeSelector` but using a more expressive syntax), while the latter specifies *preferences* that the scheduler
+will try to enforce but will not guarantee. The "IgnoredDuringExecution" part of the names means that, similar
+to how `nodeSelector` works, if labels on a node change at runtime such that the rules on a pod are no longer
+met, the pod will still continue to run on the node. In the future we plan to offer
+`requiredDuringSchedulingRequiredDuringExecution` which will be just like `requiredDuringSchedulingIgnoredDuringExecution`
+except that it will evict pods from nodes that cease to satisfy the pods' node affinity requirements.
+
+Node affinity is currently expressed using an annotation on Pod. In v1.3 it will use a field, and we will
+also introduce the second piece of the affinity mechanism, called [pod affinity](../../design/podaffinity.md),
+which allows you to control whether a pod schedules onto a particular node based on which other pods are
+running on the node, rather than the labels on the node.
+
+Here's an example of a pod that uses node affinity:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+	name: with-labels
+	annotations:
+	scheduler.alpha.kubernetes.io/affinity: >
+	{
+		"nodeAffinity": {
+			"requiredDuringSchedulingIgnoredDuringExecution": {
+				"nodeSelectorTerms": [
+				{
+					"matchExpressions": [
+					{
+						"key": "kubernetes.io/e2e-az-name",
+						"operator": "In",
+						"values": ["e2e-az1", "e2e-az2"]
+					}
+					]
+				}
+				]
+			},
+			"preferredDuringSchedulingIgnoredDuringExecution": [
+				{
+					"weight": 10,
+					"preference": {"matchExpressions": [
+						{
+							"key": "foo",
+							"operator": "In", "values": ["bar"]
+						}
+					]}
+				}
+			]
+		}
+	}
+    another-annotation-key: another-annotation-value
+spec:
+  containers:
+  - name: with-labels
+    image: gcr.io/google_containers/pause:2.0
+```
+
+This node affinity rule says the pod can only be placed on a node with a label whose key is
+`kubernetes.io/e2e-az-name` and whose value is either `e2e-az1` or `e2e-az2`. In addition,
+among nodes that meet that criteria, nodes with a label whose key is `foo` and whose
+value is `bar` should be preferred.
+
+If you specify both `nodeSelector` and `nodeAffinity`, *both* must be satisfied for the pod
+to be scheduled onto a candidate node.
+
+### Built-in node labels
+
+In addition to labels you [attach yourself](#step-one-attach-label-to-the-node), nodes come pre-populated
+with a standard set of labels. As of Kubernetes v1.2 these labels are
+* `hostname`
```

Comment:
`kubernetes.io/hostname`:
https://github.com/kubernetes/kubernetes/blob/e92f5890aafefba3934c63524b74260954c3d2c2/pkg/kubelet/kubelet.go#L973

That's unfortunate. At least it is actually the hostname rather than the node name, but it's still confusing.